README.rst
====================================
Python bindings for SpinRewriter API
====================================

`Spin Rewriter <http://www.spinrewriter.com/bigcontentsearch>`_ is an online
service for spinning text (synonym substitution) that creates unique version(s)
of existing text. This package provides a way to easily interact with
`SpinRewriter API <http://www.spinrewriter.com/api>`_. Usage requires an
account, `get one here <http://www.spinrewriter.com/registration>`_.

* `Source code @ GitHub <https://github.com/niteoweb/spinrewriter>`_
* `Releases @ PyPI <http://pypi.python.org/pypi/spinrewriter#downloads>`_
* `Documentation @ ReadTheDocs <http://readthedocs.org/docs/spinrewriter>`_
* `Continuous Integration @ Travis-CI <http://travis-ci.org/niteoweb/spinrewriter>`_


Install
=======

Install into your Python path using `pip` or `easy_install`::

    $ pip install spinrewriter
    $ easy_install spinrewriter


Usage
=====

After installing it, this is how you use it::

    Initialize SpinRewriter.
    >>> text = u"This is the text we want to spin."
    >>> from spinrewriter import SpinRewriter
    >>> rewriter = SpinRewriter('username', 'api_key')

    Request processed spun text with spintax.
    >>> rewriter.text_with_spintax(text)
    u"{This is|This really is|That is|This can be} some text that we'd {like to
    |prefer to|want to|love to} spin."

    Request a unique variation of processed given text.
    >>> rewriter.unique_variation(text)
    u"This really is some text that we'd love to spin."




docs/HISTORY.rst
Changelog
=========

0.1.6 (unreleased)
------------------

- Replace development tools and 100% test coverage.
  [matejc]


0.1.5 (2012-12-17)
------------------

- Added tests for parsing error messages and fixed loads of nasty bugs found
  while writing these tests.
  [zupo]


0.1.4 (2012-11-06)
------------------

- Fixed spelling error in one of authentication error's message.
  [zupo]

- Use 'vvv' for syntax validation.
  [matejc]


0.1.3 (2012-07-31)
------------------

- Fixed unicode encode/decode error for article texts containing non-ascii
  characters.
  [plamut]

- Added missing test coverage for the unique_variation_from_spintax method.
  [plamut]


0.1.2 (2012-07-24)
------------------

- Various fixes of bugs that surfaced when lib was put into staging.
  [zupo]


0.1.1 (2012-04-13)
------------------

- A URL in README.txt was missing a leading ``http`` which broke reST rendering
  on PyPI.
  [zupo]


0.1 (2012-04-13)
----------------

- SpinRewriter facade class.
  [plamut]

- Tests and documentation.
  [zupo]

- Raw API access class.
  [plamut]

- Project skeleton.
  [zupo]



src/spinrewriter/tests.py
# -*- coding: utf-8 -*-
from spinrewriter import Api

import unittest2 as unittest
import mock


class TestApi(unittest.TestCase):

    def setUp(self):
        """Utility code shared among all tests."""
        self.api = Api('foo@bar.com', 'test_api_key')

    def test_init(self):
        """Test that __init__ correctly stores initialization value.

        self.api has already been initialized in setUp() so we just need to
        check for values.
        """
        self.assertEquals(self.api.email_address, 'foo@bar.com')
        self.assertEquals(self.api.api_key, 'test_api_key')

    @mock.patch('spinrewriter.urllib2')
    @mock.patch('spinrewriter.urllib')
    def test_send_request(self, urllib, urllib2):
        """Test that _send_requests correctly parses JSON response into a dict
        and that request parameters get encoded beforehand.
        """
        # mock response from connection
        urllib2.urlopen.return_value.read.return_value = '{"foo":"bar"}'

        # call it
        result = self.api._send_request({'foo': 'bar'})

        # test response
        self.assertEquals(result['foo'], 'bar')

        # were parameters encoded?
        urllib.urlencode.assert_called_with({'foo': 'bar'})

    @mock.patch('spinrewriter.urllib2')
    def test_api_quota_call(self, urllib2):
        """Test if Api.api_quota() correctly parses the response it gets from
        SpinRewriter API.
        """

        # mock response from urllib2
        mocked_response = u"""{
            "status":"OK",
            "response":"You made 19 API requests in the last 24 hours. 81 still available.",
            "api_requests_made":19,"api_requests_available":81
        }"""
        urllib2.urlopen.return_value.read.return_value = mocked_response

        # call API
        result = self.api.api_quota()

        # test responses
        self.assertEquals(result['status'], u'OK')
        self.assertEquals(result['api_requests_made'], 19)
        self.assertEquals(result['api_requests_available'], 81)
        self.assertEquals(result['response'], u'You made 19 API requests in the last 24 hours. 81 still available.')

    @mock.patch('spinrewriter.urllib2')
    def test_text_with_spintax_call(self, urllib2):
        """Test if Api.text_with_spintax() correctly parses the response it gets
        from SpinRewriter API.
        """

        # mock response from urllib2
        mocked_response = u"""{
            "status":"OK",
            "response":"This is my {pet|animal|dog|cat} food.",
            "api_requests_made":20,
            "api_requests_available":80,
            "protected_terms":"food, cat",
            "nested_spintax":"false",
            "confidence_level":"medium"
        }"""
        urllib2.urlopen.return_value.read.return_value = mocked_response

        # call API
        result = self.api.text_with_spintax(
            text="This is my pet food.",
            protected_terms=['food', 'cat']
        )

        # test results
        self.assertEquals(result['status'], u'OK')
        self.assertEquals(result['api_requests_made'], 20)
        self.assertEquals(result['api_requests_available'], 80)
        self.assertEquals(result['protected_terms'], u"food, cat")
        self.assertEquals(result['nested_spintax'], u"false")
        self.assertEquals(result['confidence_level'], u"medium")
        self.assertEquals(result['response'], u"This is my {pet|animal|dog|cat} food.")

    @mock.patch('spinrewriter.urllib2')
    def test_unique_variation_call(self, urllib2):
        """Test if Api.unique_variation() correctly parses the response it gets
        from SpinRewriter API.
        """

        # mock response from urllib2
        mocked_response = u"""{
            "status":"OK",
            "response":"This is my animal food.",
            "api_requests_made":21,
            "api_requests_available":79,
            "protected_terms":"food, cat",
            "nested_spintax":"false",
            "confidence_level":"medium"
        }"""
        urllib2.urlopen.return_value.read.return_value = mocked_response

        # call API
        result = self.api.unique_variation(
            text="This is my pet food.",
            protected_terms=['food', 'cat']
        )

        # test results
        self.assertEquals(result['status'], u'OK')
        self.assertEquals(result['api_requests_made'], 21)
        self.assertEquals(result['api_requests_available'], 79)
        self.assertEquals(result['protected_terms'], u"food, cat")
        self.assertEquals(result['nested_spintax'], u"false")
        self.assertEquals(result['confidence_level'], u"medium")
        self.assertEquals(result['response'], u"This is my animal food.")

    @mock.patch('spinrewriter.urllib2')
    def test_transform_plain_text_call(self, urllib2):
        """Test if Api.transform_plain_text() correctly parses the response it
        gets from SpinRewriter API. This method is used by unique_variation()
        and text_with_spintax().
        """

        # mock response from urllib2
        mocked_response = u"""{
            "status":"OK",
            "response":"This is my animal food.",
            "api_requests_made":22,
            "api_requests_available":78,
            "protected_terms":"",
            "nested_spintax":"false",
            "confidence_level":"medium"
        }"""
        urllib2.urlopen.return_value.read.return_value = mocked_response

        # call API
        result = self.api._transform_plain_text(
            action=Api.ACTION.unique_variation,
            text="This is my pet food.",
            protected_terms=[],
            confidence_level=Api.CONFIDENCE_LVL.medium,
            nested_spintax=False,
            spintax_format=Api.SPINTAX_FORMAT.pipe_curly,
        )

        # test results
        self.assertEquals(result['status'], u'OK')
        self.assertEquals(result['api_requests_made'], 22)
        self.assertEquals(result['api_requests_available'], 78)
        self.assertEquals(result['protected_terms'], u"")
        self.assertEquals(result['nested_spintax'], u"false")
        self.assertEquals(result['confidence_level'], u"medium")
        self.assertEquals(result['response'], u"This is my animal food.")

    @mock.patch('spinrewriter.Api._send_request')
    def test_protected_terms_transformation(self, _send_request):
        """Test that protected_terms are correctly transformed into a string."""
        # prepare arguments for calling _transform_plain_text
        args = dict(
            action=Api.ACTION.unique_variation,
            text="This is my pet food.",
            protected_terms=['food', 'cat'],
            confidence_level=Api.CONFIDENCE_LVL.medium,
            nested_spintax=False,
            spintax_format=Api.SPINTAX_FORMAT.pipe_curly,
        )

        # we don't care what the response is
        _send_request.return_value = None

        # call it
        self.api._transform_plain_text(**args)

        # now test that protected_terms are in correct format
        _send_request.assert_called_with((
            ('email_address', 'foo@bar.com'),
            ('api_key', 'test_api_key'),
            ('action', 'unique_variation'),
            ('text', 'This is my pet food.'),
            ('protected_terms', 'food\ncat'),  # This is the only line we are interested in here, it needs to be newline-separated
            ('confidence_level', 'medium'),
            ('nested_spintax', False),
            ('spintax_format', '{|}'),
        ))

    @mock.patch('spinrewriter.Api._send_request')
    def test_protected_terms_empty(self, _send_request):
        """Test that correct default value is set for protected_terms if the
        list is empty.
        """
        # prepare arguments for calling _transform_plain_text
        args = dict(
            action=Api.ACTION.unique_variation,
            text="This is my pet food.",
            protected_terms=[],
            confidence_level=Api.CONFIDENCE_LVL.medium,
            nested_spintax=False,
            spintax_format=Api.SPINTAX_FORMAT.pipe_curly,
        )

        # we don't care what the response is
        _send_request.return_value = None

        # call it
        self.api._transform_plain_text(**args)

        # now test that protected_terms are in correct format
        _send_request.assert_called_with((
            ('email_address', 'foo@bar.com'),
            ('api_key', 'test_api_key'),
            ('action', 'unique_variation'),
            ('text', 'This is my pet food.'),
            ('protected_terms', ''),  # This is the only line we are interested in here, it needs to be an empty string, not an empty list
            ('confidence_level', 'medium'),
            ('nested_spintax', False),
            ('spintax_format', '{|}'),
        ))


src/spinrewriter/tests/test_errors.py
# -*- coding: utf-8 -*-
from spinrewriter import Api
from spinrewriter import exceptions as ex
import unittest2 as unittest


class TestErrors(unittest.TestCase):

    def setUp(self):
        """Utility code shared among all tests."""
        self.api = Api('foo@bar.com', 'test_api_key')

    def test_parsing_error_messages(self):
        """Go through all possible error messages and see that they are parsed
        correctly.
        """

        msg = u'Authentication with Spin Rewriter API failed.'
        with self.assertRaises(ex.AuthenticationError) as cm:
            self.api._raise_error(
                {'response': 'Authentication failed. Unique API key is not '
                             'valid for this user.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.AuthenticationError) as cm:
            self.api._raise_error(
                {'response': 'Authentication failed. No user with this email '
                             'address found.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.AuthenticationError) as cm:
            self.api._raise_error(
                {'response': 'This user does not have a valid Spin Rewriter '
                             'subscription.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Quota limit for API calls reached.'
        with self.assertRaises(ex.QuotaLimitError) as cm:
            self.api._raise_error(
                {'response': 'API quota exceeded. You can make 50 requests '
                             'per day.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Not enough time passed since last API request.'
        with self.assertRaises(ex.UsageFrequencyError) as cm:
            self.api._raise_error(
                {'response': 'You can only submit entirely new text for '
                             'analysis once every 5 seconds.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Unknown API action requested.'
        with self.assertRaises(ex.UnknownActionError) as cm:
            self.api._raise_error(
                {'response': 'Requested action does not exist. Please refer '
                             'to the Spin Rewriter API documentation.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Required parameter not present in API request.'
        with self.assertRaises(ex.MissingParameterError) as cm:
            self.api._raise_error(
                {'response': 'Email address and unique API key are both '
                             'required. At least one is missing.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Invalid parameter value passed to API.'
        with self.assertRaises(ex.ParamValueError) as cm:
            self.api._raise_error(
                {'response': 'Original text too short.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.ParamValueError) as cm:
            self.api._raise_error(
                {'response': 'Original text too long. '
                             'Text can have up to 4,000 words.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.ParamValueError) as cm:
            self.api._raise_error(
                {'response': 'Original text after analysis too long. '
                             'Text can have up to 4,000 words.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.ParamValueError) as cm:
            self.api._raise_error(
                {'response': 'Spinning syntax invalid. With this action you '
                             'should provide text with existing valid '
                             '{first option|second option} spintax.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.ParamValueError) as cm:
            self.api._raise_error(
                {'response': 'The {first|second} spinning syntax invalid. '
                             'Re-check the syntax, i.e. '
                             'curly brackets and pipes.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.ParamValueError) as cm:
            self.api._raise_error(
                {'response': 'Spinning syntax invalid.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Internal error occured on API server when processing request.'
        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'Analysis of your text failed. '
                             'Please inform us about this.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'Synonyms for your text could not be loaded. '
                             'Please inform us about this.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'Unable to load your new analyzed project.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'Unable to load your existing analyzed project.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'Unable to find your project in the database.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'Unable to load your analyzed project.'})
        self.assertEqual(msg, str(cm.exception))

        with self.assertRaises(ex.InternalApiError) as cm:
            self.api._raise_error(
                {'response': 'One-Click Rewrite failed.'})
        self.assertEqual(msg, str(cm.exception))

        msg = u'Unrecognized API error message received: foo'
        with self.assertRaises(ex.UnknownApiError) as cm:
            self.api._raise_error(
                {'response': 'foo'})
        self.assertEqual(msg, str(cm.exception))


src/spinrewriter/exceptions.py
# -*- coding: utf-8 -*-


class SpinRewriterApiError(Exception):
    """Base class for exceptions in Spin Rewriter module."""
    def __init__(self, api_error_msg):
        #api_error_msg respresents raw error string as returned by API server
        super(SpinRewriterApiError, self).__init__()
        self.api_error_msg = api_error_msg


class AuthenticationError(SpinRewriterApiError):
    """Raised when authentication error occurs."""
    def __str__(self):
        return u'Authentication with Spin Rewriter API failed.'


class QuotaLimitError(SpinRewriterApiError):
    """Raised when API quota limit is reached."""
    def __str__(self):
        return u'Quota limit for API calls reached.'


class UsageFrequencyError(SpinRewriterApiError):
    """Raised when subsequent API requests are made in a too short time
    interval."""
    def __str__(self):
        return u'Not enough time passed since last API request.'


class UnknownActionError(SpinRewriterApiError):
    """Raised when unknown API action is requested."""
    def __str__(self):
        return u'Unknown API action requested.'


class MissingParameterError(SpinRewriterApiError):
    """Raised when required parameter is not provided."""
    def __str__(self):
        return u'Required parameter not present in API request.'


class ParamValueError(SpinRewriterApiError):
    """Raised when parameter passed to API call has an invalid value."""
    def __str__(self):
        return u'Invalid parameter value passed to API.'


class InternalApiError(SpinRewriterApiError):
    """Raised when unexpected error occurs on the API server when processing a
    request."""
    def __str__(self):
        return u'Internal error occured on API server when processing request.'


class UnknownApiError(SpinRewriterApiError):
    """Raised when API call results in an unrecognized error."""
    def __str__(self):
        return (u'Unrecognized API error message received: {}'.format(
            self.api_error_msg))


src/spinrewriter/__init__.py
# -*- coding: utf-8 -*-

import json
import re
import string
import urllib
import urllib2

from collections import namedtuple

from spinrewriter import exceptions as ex


class Api(object):
    """A class representing the Spin Rewriter API
    (http://www.spinrewriter.com/).
    """
    URL = 'http://www.spinrewriter.com/action/api'
    """URL for invoking the API"""

    _tmp_list = ['api_quota', 'text_with_spintax', 'unique_variation',
                 'unique_variation_from_spintax']
    ACTION = namedtuple('ACTION', _tmp_list)(*_tmp_list)
    """collection of possible values for the action parameter"""

    _tmp_list = ['low', 'medium', 'high']
    CONFIDENCE_LVL = namedtuple('CONFIDENCE_LVL', ['low', 'medium', 'high'])(
        *_tmp_list)
    """collection of possible values for the confidence_level parameter"""

    SPINTAX_FORMAT = namedtuple(
        'SPINTAX_FORMAT',
        ['pipe_curly', 'tilde_curly', 'pipe_square', 'spin_tag']
    )(*['{|}', '{~}', '[|]', '[spin]'])
    """collection of possible values for the spintax_format parameter"""

    _tmp_list = ['email_address', 'api_key',
                 'action', 'text', 'protected_terms',
                 'confidence_level', 'nested_spintax', 'spintax_format']
    REQ_P_NAMES = namedtuple('REQ_P_NAMES', _tmp_list)(*_tmp_list)
    """collection of all request parameters' names"""

    _tmp_list = ['status', 'response', 'api_requests_made',
                 'api_requests_available', 'protected_terms',
                 'confidence_level']
    RESP_P_NAMES = namedtuple('RESP_P_NAMES', _tmp_list)(*_tmp_list)
    """collection of all response fields' names"""

    _tmp_list = ['ok', 'error']
    STATUS = namedtuple('STATUS', ['ok', 'error'])(*map(
        string.upper, _tmp_list))
    """possible response status strings returned by API"""

    def __init__(self, email_address, api_key):
        self.email_address = email_address
        self.api_key = api_key

    def api_quota(self):
        """Return the number of made and remaining API calls for the 24-hour
        period.

        :return: remaining API quota
        :rtype: dictionary
        """
        params = (
            (self.REQ_P_NAMES.email_address, self.email_address),
            (self.REQ_P_NAMES.api_key, self.api_key),
            (self.REQ_P_NAMES.action, self.ACTION.api_quota),
        )
        response = self._send_request(params)
        return response

    def text_with_spintax(self, text, protected_terms=None,
                          confidence_level=CONFIDENCE_LVL.medium,
                          nested_spintax=False,
                          spintax_format=SPINTAX_FORMAT.pipe_curly):
        """Return processed spun text with spintax.

        :param text: original text that needs to be changed
        :type text: string
        :param protected_terms: (optional) keywords and key phrases that
            should be left intact
        :type protected_terms: list of strings
        :param confidence_level: (optional) the confidence level of
            the One-Click Rewrite process
        :type confidence_level: string
        :param nested_spintax: (optional) whether or not to also spin
            single words inside already spun phrases
        :type nested_spintax: boolean
        :param spintax_format: (optional) spintax format to use in
            returned text
        :type spintax_format: string

        :return: processed text and some other meta info
        :rtype: dictionary
        """
        response = self._transform_plain_text(
            self.ACTION.text_with_spintax,
            text,
            protected_terms,
            confidence_level,
            nested_spintax,
            spintax_format
        )

        if response[self.RESP_P_NAMES.status] == self.STATUS.error:
            self._raise_error(response)
        else:
            return response

    def unique_variation(
            self,
            text,
            protected_terms=None,
            confidence_level=CONFIDENCE_LVL.medium,
            nested_spintax=False,
            spintax_format=SPINTAX_FORMAT.pipe_curly
    ):
        """Return a unique variation of the given text.

        :param text: original text that needs to be changed
        :type text: string
        :param protected_terms: (optional) keywords and key phrases that
            should be left intact
        :type protected_terms: list of strings
        :param confidence_level: (optional) the confidence level of
            the One-Click Rewrite process
        :type confidence_level: string
        :param nested_spintax: (optional) whether or not to also spin
            single words inside already spun phrases
        :type nested_spintax: boolean
        :param spintax_format: (optional) (probably not relevant here?
            But API documentation not clear here ...)
        :type spintax_format: string

        :return: processed text and some other meta info
        :rtype: dictionary
        """
        response = self._transform_plain_text(
            self.ACTION.unique_variation, text, protected_terms,
            confidence_level, nested_spintax, spintax_format
        )

        if response[self.RESP_P_NAMES.status] == self.STATUS.error:
            self._raise_error(response)
        else:
            return response

    def unique_variation_from_spintax(
            self,
            text,
            nested_spintax=False,
            spintax_format=SPINTAX_FORMAT.pipe_curly
    ):
        """Return a unique variation of an already spun text.

        :param text: text to process
        :type text: string
        :param nested_spintax: whether or not to also spin single words
            inside already spun phrases
        :type nested_spintax: boolean
        :param spintax_format: (probably not relevant here?
            But API documentation not clear here ...)
        :type spintax_format: string

        :return: processed text and some other meta info
        :rtype: dictionary
        """
        params = (
            (self.REQ_P_NAMES.email_address, self.email_address),
            (self.REQ_P_NAMES.api_key, self.api_key),
            (
                self.REQ_P_NAMES.action,
                self.ACTION.unique_variation_from_spintax
            ),
            (self.REQ_P_NAMES.text, text.encode('utf-8')),
            (self.REQ_P_NAMES.nested_spintax, nested_spintax),
            (self.REQ_P_NAMES.spintax_format, spintax_format),
        )
        response = self._send_request(params)

        if response[self.RESP_P_NAMES.status] == self.STATUS.error:
            self._raise_error(response)
        else:
            return response

    def _send_request(self, params):
        """Invoke Spin Rewriter API with given parameters and return its
        response.

        :param params: parameters to pass along with the request
        :type params: tuple of 2-tuples

        :return: API's response (already JSON-decoded)
        :rtype: dictionary
        """
        con = urllib2.urlopen(self.URL, urllib.urlencode(params))
        response = con.read()
        return json.loads(response)

    def _raise_error(self, api_response):
        """Examine the API response and raise exception of the appropriate
        type.

        NOTE: usage of this method only makes sense when API response's
        status indicates an error

        :param api_response: API's response fileds
        :type api_response: dictionary
        """
        error_msg = api_response[self.RESP_P_NAMES.response]

        if (
            re.match(
                r'Authentication failed. '
                r'No user with this email address found.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Authentication failed. '
                r'Unique API key is not valid for this user.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'This user does not have a valid Spin Rewriter subscription.',
                error_msg,
                re.IGNORECASE
            )
        ):
            raise ex.AuthenticationError(error_msg)

        elif re.match(
            r'API quota exceeded. '
            r'You can make \d+ requests per day.',
            error_msg,
            re.IGNORECASE
        ):
            raise ex.QuotaLimitError(error_msg)

        elif re.match(
            r'You can only submit entirely new text '
            r'for analysis once every \d+ seconds.',
            error_msg,
            re.IGNORECASE
        ):
            raise ex.UsageFrequencyError(error_msg)

        elif re.match(
            r'Requested action does not exist. '
            r'Please refer to the Spin Rewriter API documentation.',
            error_msg,
            re.IGNORECASE
        ):
            # NOTE: This should never occur unless
            # there is a bug in the API library.
            raise ex.UnknownActionError(error_msg)

        elif re.match(
            r'Email address and unique API key are both required. '
            r'At least one is missing.',
            error_msg,
            re.IGNORECASE
        ):
            # NOTE: This should never occur unless
            # there is a bug in the API library.
            raise ex.MissingParameterError(error_msg)

        elif (
            re.match(
                r'Original text too short.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Original text too long. Text can have up to 4,000 words.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Original text after analysis too long. '
                r'Text can have up to 4,000 words.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Spinning syntax invalid. '
                r'With this action you should provide text with existing valid'
                r'{first option|second option} spintax.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'The {first|second} spinning syntax invalid. '
                r'Re-check the syntax, i.e. curly brackets and pipes\.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Spinning syntax invalid.',
                error_msg,
                re.IGNORECASE
            )
        ):
            raise ex.ParamValueError(error_msg)

        elif (
            re.match(
                r'Analysis of your text failed. '
                r'Please inform us about this.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Synonyms for your text could not be loaded. '
                r'Please inform us about this.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Unable to load your new analyzed project.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Unable to load your existing analyzed project.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Unable to find your project in the database.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'Unable to load your analyzed project.',
                error_msg,
                re.IGNORECASE
            ) or
            re.match(
                r'One-Click Rewrite failed.',
                error_msg,
                re.IGNORECASE
            )
        ):
            raise ex.InternalApiError(error_msg)

        else:
            raise ex.UnknownApiError(error_msg)

    def _transform_plain_text(
            self,
            action,
            text,
            protected_terms,
            confidence_level,
            nested_spintax,
            spintax_format
    ):
        """Transform plain text using SpinRewriter API.

        Pack parameters into format as expected by the _send_request method and
        invoke the action method to get transformed text from the API.

        :param action: name of the action that will be requested from API
        :type action: string
        :param text: text to process
        :type text: string
        :param protected_terms: keywords and key phrases that should be left
            intact
        :type protected_terms: list of strings
        :param confidence_level: the confidence level of the One-Click Rewrite
            process
        :type confidence_level: string
        :param nested_spintax: whether or not to also spin single words inside
            already spun phrases
        :type nested_spintax: boolean
        :param spintax_format: spintax format to use in returned text
        :type spintax_format: string

        :return: processed text and some other meta info
        :rtype: dictionary
        """
        if protected_terms:
            # protected_terms could be separated by other characters too,
            # like commas
            protected_terms = [
                term.encode('utf-8') for term in protected_terms]
            protected_terms = '\n'.join(protected_terms)
        else:
            protected_terms = ''

        params = (
            (self.REQ_P_NAMES.email_address, self.email_address),
            (self.REQ_P_NAMES.api_key, self.api_key),
            (self.REQ_P_NAMES.action, action),
            (self.REQ_P_NAMES.text, text.encode('utf-8')),
            (self.REQ_P_NAMES.protected_terms, protected_terms),
            (self.REQ_P_NAMES.confidence_level, confidence_level),
            (self.REQ_P_NAMES.nested_spintax, nested_spintax),
            (self.REQ_P_NAMES.spintax_format, spintax_format),
        )
        return self._send_request(params)


class SpinRewriter(object):
    """A facade for easier usage of the raw Spin Rewriter API."""

    def __init__(self, email_address, api_key):
        self.email_address = email_address
        self.api_key = api_key
        self.api = Api(email_address, api_key)

    def unique_variation(
            self, text, confidence_level=Api.CONFIDENCE_LVL.medium):
        """Return unique variation of the given text.

        :param text: text to process
        :type text: string
        :param confidence_level: how 'confident' the spinner API is when
            transforming the text
        :type confidence_level: Api.CONFIDENCE_LVL

        :return: spinned version of the original text
        :rtype: string
        """
        response = self.api.unique_variation(text, confidence_level)
        return response[Api.RESP_P_NAMES.response]

    def text_with_spintax(
            self, text, confidence_level=Api.CONFIDENCE_LVL.medium):
        """Return text with spintax elements inserted.

        :param text: text to process
        :type text: string
        :param confidence_level: how 'confident' the spinner API is when
            transforming the text
        :type confidence_level: Api.CONFIDENCE_LVL

        :return: original text with spintax elements
        :rtype: string
        """
        response = self.api.text_with_spintax(text, confidence_level)
        return response[Api.RESP_P_NAMES.response]
